File: basic-data-structures/basic-data-structures.ts
// === Arrays ===

const names: string[] = ["Alice", "Bob", "Charlie", "Diana"];

// O(1) Access
console.log("Array Access:", names[2]);

// O(n) Search
function searchArray(arr: string[], target: string): boolean {
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === target) return true;
  }
  return false;
}
console.log("Array Search:", searchArray(names, "Bob"));

// === Hash Tables (Maps) ===

const phoneBook = new Map<string, string>();

// O(1) Average - Insertion
phoneBook.set("Alice", "555-0192");
phoneBook.set("Bob", "555-8832");

// O(1) Average - Access
console.log("Hash Table Access:", phoneBook.get("Alice"));

// === Stacks (LIFO) ===

class Stack<T> {
  private items: T[] = [];

  push(element: T): void {
    this.items.push(element); // O(1)
  }

  pop(): T | undefined {
    return this.items.pop(); // O(1)
  }

  peek(): T | undefined {
    return this.items[this.items.length - 1]; // O(1)
  }
}

const myStack = new Stack<number>();
myStack.push(10);
myStack.push(20);
console.log("Stack Peek:", myStack.peek());
console.log("Stack Pop:", myStack.pop());

// === Queues (FIFO) ===

class Queue<T> {
  private items: T[] = [];

  enqueue(element: T): void {
    this.items.push(element); // O(1)
  }

  dequeue(): T | undefined {
    // Note: Array shift is O(n). Real queues use Linked Lists for O(1) Dequeue.
    return this.items.shift();
  }

  peek(): T | undefined {
    return this.items[0]; // O(1)
  }
}

const myQueue = new Queue<string>();
myQueue.enqueue("Alice");
myQueue.enqueue("Bob");
console.log("Queue Peek:", myQueue.peek());
console.log("Queue Dequeue:", myQueue.dequeue());

File: basic-data-structures/basic-data-structures.go
package main

import "fmt"

func main() {
	// === Arrays (Slices in Go) ===

	names := []string{"Alice", "Bob", "Charlie", "Diana"}

	// O(1) Access
	fmt.Println("Array Access:", names[2])

	// O(n) Search
	searchArray := func(arr []string, target string) bool {
		for _, item := range arr {
			if item == target {
				return true
			}
		}
		return false
	}
	fmt.Println("Array Search:", searchArray(names, "Bob"))

	// === Hash Tables (Maps in Go) ===

	phoneBook := make(map[string]string)

	// O(1) Average - Insertion
	phoneBook["Alice"] = "555-0192"
	phoneBook["Bob"] = "555-8832"

	// O(1) Average - Access
	fmt.Println("Hash Table Access:", phoneBook["Alice"])


	// === Stacks (LIFO) ===
	
	stack := []int{}
	stack = append(stack, 10) // Push O(1)
	stack = append(stack, 20)

	fmt.Println("Stack Peek:", stack[len(stack)-1]) // Peek O(1)
	
	// Pop O(1)
	poppedValue := stack[len(stack)-1]
	stack = stack[:len(stack)-1]
	fmt.Println("Stack Pop:", poppedValue)


	// === Queues (FIFO) ===
	
	queue := []string{}
	queue = append(queue, "Alice") // Enqueue O(1)
	queue = append(queue, "Bob")

	fmt.Println("Queue Peek:", queue[0]) // Peek O(1)

	// Dequeue (In Go slices this reorganizes the underlying array or shifts pointers)
	dequeuedValue := queue[0]
	queue = queue[1:]
	fmt.Println("Queue Dequeue:", dequeuedValue)
}
